lib/oauthdb.js

```diff
@@ -0,0 +1,138 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+/* Operations on OAuth database state.
+ *
+ * Currently this is not actually talking to a database,
+ * it's making authenticated calls to the fxa-oauth-server API
+ * to interrogate and manipulate its state, essentially treating
+ * fxa-oauth-server as a kind of backend micro-service.
+ *
+ * We want to work towards merging the fxa-oauth-server code
+ * directly into the main fxa-auth-server process, at which point
+ * this abstraction will convert into more direct db access.
+ *
+ */
+
+const Joi = require('joi')
+const JWT = require('jsonwebtoken')
+const P = require('./promise')
+
+const Pool = require('./pool')
+const {
+  HEX_STRING,
+  DISPLAY_SAFE_UNICODE
+} = require('./routes/validators')
+
+function objMap(obj, func) {
+  return Object.keys(obj).reduce((acc, k) => {
+    return Object.assign(acc, { [k]: func(obj[k]) })
+  }, {})
+}
+
+module.exports = (log, config) => {
+
+  const SafeUrl = require('./safe-url')(log)
+
+  const pool = new Pool(config.oauth.url, config.oauth.poolee)
+
+  const SAFE_URLS = {}
+
+  SAFE_URLS.getClientInfo = new SafeUrl('/v1/client/:clientId', 'oauthdb.getClientInfo', Joi.object({
+    id: Joi.string().length(16).regex(HEX_STRING).required(),
+    name: Joi.string().max(25).regex(DISPLAY_SAFE_UNICODE).required(),
+    trusted: Joi.boolean().required(),
+    image_uri: Joi.any(),
+    redirect_uri: Joi.string().rqeuired().allow('')
+  }))
+  SAFE_URLS.getScopedKeyData = new SafeUrl('/v1/key-data', 'oauthdb.getScopedKeyData', Joi.object().pattern(/^/, [
+    Joi.object({
+      identifier: Joi.string.required(),
+      keyRotationSecret: Joi.string.required(),
+      keyRotationTimestamp: Joi.number().required(),
+    })
+  ]))
```

Comment:
I wanted to validate the responses here, and adding that functionality to our `SafeUrl` class seemed like a neat way to do it in a kind of declarative fashion.  If we like the abstraction here, I think we could also use it for e..g the pushbox integration code.